bqemulatormanager/schema.py

```diff
@@ -1,48 +1,69 @@
 import os
-from typing import List
+from typing import List, Optional, Protocol
 
 import yaml
 from google.cloud import bigquery
+from google.cloud.bigquery.table import Table
+
+
+class SchemaError(Exception):
+    pass
+
+
+class BigQueryClient(Protocol):
+    def get_table(self, _table: str) -> Table:  # noqa: U101
+        pass
 
 
 class SchemaManager:
-    def __init__(self, master_path: str = "master_schema.yaml", client=None):
-        self.client = client
-        self.master_path = master_path
-        self.change_flg = False
-        if os.path.isfile(master_path):
-            with open(master_path) as f:
+    """
+    SchemaManager manage schema of tables in emulator.
+
+    params:
+        schema_file_path:
+            path to yaml file which contains schema of tables in emulator.
+
+        client:
+            client of production BigQuery. If you set this, SchemaManager will fetch schema from production BigQuery.
+    """
+
+    def __init__(self, schema_file_path: Optional[str] = "bqem_master_schema.yaml", client: Optional[BigQueryClient] = None):
+        self.__client = client
+        self.__schema_file_path = schema_file_path
+        self.__change_flg = False
+        if schema_file_path is not None and os.path.isfile(schema_file_path):
+            with open(schema_file_path) as f:
                 master_schema = yaml.safe_load(f)
         else:
             master_schema = {}
-        self.master_schema = master_schema
-
-    def get_schema(self, table_id: str) -> List[bigquery.SchemaField]:
-        project, dataset, table = table_id.split(".")
+        self.__master_schema = master_schema
 
-        schema = self.master_schema.get(project, {}).get(dataset, {}).get(table, {})
+    def get_schema(self, project_id: str, dataset_id: str, table_id: str) -> List[bigquery.SchemaField]:
+        schema = self.__master_schema.get(project_id, {}).get(dataset_id, {}).get(table_id, {})
 
         if schema == {}:
-            schema = self._get_schema_from_production(table_id)
-            deepupdate(self.master_schema, {project: {dataset: {table: [s._properties for s in schema]}}})
-            self.change_flg = True
+            schema = self._get_schema_from_production(project_id, dataset_id, table_id)
+            if schema == []:
+                raise SchemaError(f"schema for {project_id}.{dataset_id}.{table_id} is not found in master schema and production environment")
```

Comment:
[nits][imo]

`_get_schema_from_production()` should raise this error, rather than return `[]` ?